src/Telemetry/AzureDevOpsTimeline/AzureDevOpsTimeline.cs

```diff
@@ -176,15 +180,46 @@ private async Task RunLoop(CancellationToken cancellationToken)
                     build => azureServer.GetTimelineAsync(project, build.Id, cancellationToken)
                 );
 
+            // for each timeline, look for previousAttempts.timelineId timelines
+            // For each timeline, look at timeline.lastChangedOn for ingestion cutoff
+            // timeline.lastChangedOn will be before build.finishTime
+
             await Task.WhenAll(tasks.Select(s => s.Value));
+
+            // Look for retried timelines
+            List<(Build build, Task<Timeline> timelineTask)> retriedTimelineTasks = new List<(Build, Task<Timeline>)>();
+
+            foreach ((Build build, Task<Timeline> timelineTask) in tasks)
+            {
+                Timeline timeline = await timelineTask;
+
+                IEnumerable<string> additionalTimelineIds = timeline.Records
+                    .SelectMany(record => record.PreviousAttempts)
+                    .Where(attempt => !(attempt is null))
+                    .Select(attempt => attempt.TimelineId)
+                    .Distinct(StringComparer.OrdinalIgnoreCase);
+
+                retriedTimelineTasks.AddRange(
+                    additionalTimelineIds.Select(
+                        timelineId => (build, azureServer.GetTimelineAsync(project, build.Id, timelineId, cancellationToken))));
+            }
+
+            await Task.WhenAll(retriedTimelineTasks.Select(o => o.timelineTask));
+
+            // Only get timelines happening after the cutoff
+            List<(Build build, Task<Timeline> timeline)> allNewTimelines = new List<(Build build, Task<Timeline> timeline)>();
+            allNewTimelines.AddRange(tasks.Select(t => (t.Key, t.Value)));
+            allNewTimelines.AddRange(retriedTimelineTasks
+                .Where(t => t.timelineTask.Result.LastChangedOn > latest));
```

Comment:
Hrm... The "Result" call here is a little scary.  It's fine right now, because we awaited these all already up a bit, but code moves... _But_ I also don't have an elegant solution. :-(

Reply:
Yes, I'm also not super happy with this, but went with it since everything is awaited. The only alternative I could come with was to separate the Task creation such that the waiting happens all in one place, then regular, non-Task Timeline objects get passed for processing. But this seemed like more refactoring effort than needed for this change.